### PR TITLE
CRNS-321: Update message references for deleted or updated users

### PR DESCRIPTION
**Pull request by vishalnarkhede**
## CLA

- [ ] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required).
- [ ] Code changes are tested

## Changelog

- Delete user messages when you receive `user.deleted` event
- Update `message.user` object, when `user.updated` event is received

### CODE REVIEW

File: src/client.ts [hunk 1]
```diff
@@ -1169,6 +1169,131 @@ export class StreamChat<
...
+    const refMap = this.state.userChannelReferences[user.id] || {};
+
+    for (const channelID in refMap) {
+      const channel = this.activeChannels[channelID];
+      const state = channel.state;
+
+      /** update the messages from this user. */
+      if (state) {
+        state?.updateUserMessages(user);
+      }
```
Comment (mahboubii) [inline src/client.ts:1217]:
```suggestion
      /** update the messages from this user. */
      channel.state?.updateUserMessages(user);
```

File: src/channel_state.ts [hunk 2]
```diff
@@ -603,6 +614,153 @@ export class ChannelState<
...
+          MessageType,
+          ReactionType,
+          UserType
+        >['formatMessage']
+      >
+    >,
+    user: UserResponse<UserType>,
+    hardDelete = false,
+  ) =>
+    messages.map((m) => {
```
Comment (mahboubii) [inline src/channel_state.ts:695]: for loop is way faster in this case

File: src/channel_state.ts [hunk 3]
```diff
@@ -603,6 +614,153 @@ export class ChannelState<
...
+          latest_reactions: [],
+          mentioned_users: [],
+          own_reactions: [],
+          parent_id: m.parent_id,
+          reply_count: m.reply_count,
+          status: m.status,
+          thread_participants: m.thread_participants,
+          type: 'deleted',
+          updated_at: m.updated_at,
+          user: m.user,
```
Comment (mahboubii) [inline src/channel_state.ts:719]: does it make sense to extract this into a default empty msg or sth?

File: src/channel_state.ts [hunk 4]
```diff
@@ -603,6 +614,153 @@ export class ChannelState<
...
+            CommandType,
+            EventType,
+            MessageType,
+            ReactionType,
+            UserType
+          >['formatMessage']
+        >;
+      } else {
+        return {
+          ...m,
```
Comment (mahboubii) [inline src/channel_state.ts:733]: what about the `user` attribute?

File: src/channel_state.ts [hunk 5]
```diff
@@ -603,6 +614,153 @@ export class ChannelState<
...
+  /**
+   * @private
+   *
+   * Updates the `message.user` property on messages from given user,
+   * and returns the updates messages array
+   *
+   * @param {Array<ReturnType<ChannelState<AttachmentType, ChannelType, CommandType, EventType, MessageType, ReactionType, UserType>['formatMessage']>>} messages
+   * @param {UserResponse<UserType>} user
+   */
+  _updateUserMessages = (
```
Comment (mahboubii) [inline src/channel_state.ts:627]: this can be an inline function within the `updateUserMessages` method

**Review by mahboubii — COMMENTED** [review]
Looks good
so much stuff is going on with this one, would be better to have some unit tests as well

File: src/channel_state.ts (hunk 4 above)
Comment (vishalnarkhede) [reply src/channel_state.ts:733]: It will be part of `...m`

File: src/channel_state.ts (hunk 3 above)
Comment (vishalnarkhede) [reply src/channel_state.ts:719]: I did think about it as well, but it's not exactly empty message. There are certain properties that we retain and some we remove. So didn't make sense to me to extract it out.

File: src/channel_state.ts (hunk 4 above)
Comment (mahboubii) [reply src/channel_state.ts:733]: will it retain the old user data?

Comment (vishalnarkhede) [reply src/channel_state.ts:733]: Yeah, its supposed to!! Sounds wierd, but thats how backend behaves right now!

**Review by mahboubii — APPROVED** [review]

**Pull request merged**